<a href="https://colab.research.google.com/github/theysee/sberbank_reboot/blob/master/%D0%9F%D0%BE%D1%82%D0%BE%D0%BA_3_%D0%9F%D1%80%D0%B0%D0%BA%D1%82%D0%B8%D0%BA%D0%B0_%D0%BA_%D0%B7%D0%B0%D0%BD%D1%8F%D1%82%D0%B8%D1%8E_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Практика



*   [Презентация](https://docs.google.com/presentation/d/1dLjCBWeGlhzU3A_JNg09ty-WPiAe4XObZBENp4tgC2Y/edit?usp=sharing)
*   [master](https://colab.research.google.com/drive/1TVa3TRpGfhQa3f6eKxAxmBEHsPHUVzxt)
*   [Дамп базы](https://drive.google.com/open?id=1RJx-1kh8az1OSQqnsDjCTWdzJBlDkjcB)



![схема базы](http://bdis.umeta.ru/db/db_course/labs/SQLpHTM/king.gif)

In [2]:
!pip install PyMySQL

     |████████████████████████████████| 51kB 2.5MB/s 


In [3]:
!pip install pandas

In [0]:
import pymysql
import pandas as pd
# доступны student-ы от 0 до 49

DB_HOST = '89.223.95.235'
DB_USER = 'student_12'
DB_USER_PASSWORD = 'student_12'
DB_NAME = 'sample_db'

In [11]:
conn = pymysql.connect(DB_HOST, DB_USER, DB_USER_PASSWORD, DB_NAME)
pd.read_sql_query("select * from customer", conn).head()

customer_id  ...                                           comments
0          100  ...  very friendly people to work with -- sales rep...
1          101  ...  rep called 5/8 about change in order - contact...
2          102  ...  company doing heavy promotion beginning 10/89....
3          103  ...      contact rep about new line of tennis rackets.
4          104  ...  customer with high market share (23%) due to a...

[5 rows x 11 columns]

### Задание 1
Выбрать всю информацию о каждом отделе и месте его расположения.

In [6]:
pd.read_sql_query("""
  SELECT
      department_id,
      name,
      regional_group,
      location.location_id
  FROM
      department
      INNER JOIN
      location
      ON department.location_id = location.location_id
""", conn)

department_id        name regional_group  location_id
0              10  accounting       new york          122
1              12    research       new york          122
2              13       sales       new york          122
3              14  operations       new york          122
4              20    research         dallas          124
5              23       sales         dallas          124
6              24  operations         dallas          124
7              30       sales        chicago          123
8              34  operations        chicago          123
9              40  operations         boston          167
10             43       sales         boston          167

### Задание 2
Выбрать список фамилий сотрудников по городам, где они работают, с указанием для каждого его должности.

In [7]:
# шаг 1: соединили таблицы employee и job
pd.read_sql_query("""
  SELECT
    *
  FROM
    job
    INNER JOIN
    employee
    ON job.job_id = employee.job_id
""", conn)
# Должность есть
# Теперь надо вывести города, где они работают

job_id     function  employee_id  ...  salary commission department_id
0      667        clerk         7369  ...   800.0        NaN            20
1      667        clerk         7876  ...  1100.0        NaN            20
2      667        clerk         7900  ...   950.0        NaN            30
3      667        clerk         7919  ...   800.0        NaN            12
4      667        clerk         7934  ...  1300.0        NaN            10
5      667        clerk         7950  ...   750.0        NaN            13
6      667        clerk         7954  ...   750.0        NaN            23
7      668        staff         7609  ...  1800.0        NaN            24
8      668        staff         7676  ...  1850.0        NaN            34
9      669      analyst         7788  ...  3000.0        NaN            20
10     669      analyst         7799  ...  3000.0        NaN            12
11     669      analyst         7902  ...  3000.0        NaN            20
12     669      analyst         7916  ...  2875.0        NaN            12
13     670  salesperson         7499  ...  1600.0      300.0            30
14     670  salesperson         7521  ...  1250.0      500.0            30
15     670  salesperson         7555  ...  1250.0      300.0            13
16     670  salesperson         7557  ...  1250.0     1200.0            13
17     670  salesperson         7560  ...  1250.0        NaN            23
18     670  salesperson         7564  ...  1250.0      300.0            23
19     670  salesperson         7600  ...  1250.0      900.0            13
20     670  salesperson         7654  ...  1250.0     1400.0            30
21     670  salesperson         7789  ...  1500.0     1000.0            23
22     670  salesperson         7820  ...  1300.0      800.0            43
23     670  salesperson         7844  ...  1500.0        0.0            30
24     671      manager         7505  ...  2850.0        NaN            13
25     671      manager         7506  ...  2750.0        NaN            23
26     671      manager         7507  ...  2200.0        NaN            14
27     671      manager         7566  ...  2975.0        NaN            20
28     671      manager         7569  ...  3000.0        NaN            12
29     671      manager         7698  ...  2850.0        NaN            30
30     671      manager         7782  ...  2450.0        NaN            10
31     672    president         7839  ...  5000.0        NaN            10

[32 rows x 12 columns]

In [8]:
# шаг 2
pd.read_sql_query("""
SELECT
  *
FROM
  (SELECT
    employee_id,
    department_id,
    function
  FROM
    job
    INNER JOIN
    employee
    ON job.job_id = employee.job_id) AS job_joined
  INNER JOIN
  department
  ON job_joined.department_id = department.department_id
""", conn)
# теперь  нужны города этих департаментов

employee_id  department_id  ...        name  location_id
0          7369             20  ...    research          124
1          7876             20  ...    research          124
2          7900             30  ...       sales          123
3          7919             12  ...    research          122
4          7934             10  ...  accounting          122
5          7950             13  ...       sales          122
6          7954             23  ...       sales          124
7          7609             24  ...  operations          124
8          7676             34  ...  operations          123
9          7788             20  ...    research          124
10         7799             12  ...    research          122
11         7902             20  ...    research          124
12         7916             12  ...    research          122
13         7499             30  ...       sales          123
14         7521             30  ...       sales          123
15         7555             13  ...       sales          122
16         7557             13  ...       sales          122
17         7560             23  ...       sales          124
18         7564             23  ...       sales          124
19         7600             13  ...       sales          122
20         7654             30  ...       sales          123
21         7789             23  ...       sales          124
22         7820             43  ...       sales          167
23         7844             30  ...       sales          123
24         7505             13  ...       sales          122
25         7506             23  ...       sales          124
26         7507             14  ...  operations          122
27         7566             20  ...    research          124
28         7569             12  ...    research          122
29         7698             30  ...       sales          123
30         7782             10  ...  accounting          122
31         7839             10  ...  accounting          122

[32 rows x 6 columns]

In [9]:
# шаг 3. Соединияем с таблицей location, и формируем окончательный результат
pd.read_sql_query("""
SELECT
  employee_id,
  function 'должность',
  name as 'Имя отдела',
  regional_group 'город'
FROM
  (SELECT
    employee_id,
    job_joined.department_id,
    function,
    department.name,
    location_id
  FROM
    (SELECT
      employee_id,
      department_id,
      function
    FROM
      job
      INNER JOIN
      employee
      ON job.job_id = employee.job_id) AS job_joined
    INNER JOIN
    department
    ON job_joined.department_id = department.department_id) as tbl1
  INNER JOIN
  location
  ON tbl1.location_id = location.location_id
""", conn)

employee_id    должность  Имя отдела     город
0          7782      manager  accounting  new york
1          7839    president  accounting  new york
2          7934        clerk  accounting  new york
3          7569      manager    research  new york
4          7799      analyst    research  new york
5          7916      analyst    research  new york
6          7919        clerk    research  new york
7          7505      manager       sales  new york
8          7555  salesperson       sales  new york
9          7557  salesperson       sales  new york
10         7600  salesperson       sales  new york
11         7950        clerk       sales  new york
12         7507      manager  operations  new york
13         7369        clerk    research    dallas
14         7566      manager    research    dallas
15         7788      analyst    research    dallas
16         7876        clerk    research    dallas
17         7902      analyst    research    dallas
18         7506      manager       sales    dallas
19         7560  salesperson       sales    dallas
20         7564  salesperson       sales    dallas
21         7789  salesperson       sales    dallas
22         7954        clerk       sales    dallas
23         7609        staff  operations    dallas
24         7499  salesperson       sales   chicago
25         7521  salesperson       sales   chicago
26         7654  salesperson       sales   chicago
27         7698      manager       sales   chicago
28         7844  salesperson       sales   chicago
29         7900        clerk       sales   chicago
30         7676        staff  operations   chicago
31         7820  salesperson       sales    boston

In [10]:
# вариант 2
pd.read_sql_query("""
  SELECT
    employee_id,
    function 'должность',
    name as 'Имя отдела',
    regional_group 'город'
  FROM
    employee
    INNER JOIN
    job
    INNER JOIN
    department
    INNER JOIN
    location
  ON employee.job_id = job.job_id AND employee.department_id = department.department_id AND department.location_id = location.location_id
""", conn)

employee_id    должность  Имя отдела     город
0          7782      manager  accounting  new york
1          7839    president  accounting  new york
2          7934        clerk  accounting  new york
3          7569      manager    research  new york
4          7799      analyst    research  new york
5          7916      analyst    research  new york
6          7919        clerk    research  new york
7          7505      manager       sales  new york
8          7555  salesperson       sales  new york
9          7557  salesperson       sales  new york
10         7600  salesperson       sales  new york
11         7950        clerk       sales  new york
12         7507      manager  operations  new york
13         7369        clerk    research    dallas
14         7566      manager    research    dallas
15         7788      analyst    research    dallas
16         7876        clerk    research    dallas
17         7902      analyst    research    dallas
18         7506      manager       sales    dallas
19         7560  salesperson       sales    dallas
20         7564  salesperson       sales    dallas
21         7789  salesperson       sales    dallas
22         7954        clerk       sales    dallas
23         7609        staff  operations    dallas
24         7499  salesperson       sales   chicago
25         7521  salesperson       sales   chicago
26         7654  salesperson       sales   chicago
27         7698      manager       sales   chicago
28         7844  salesperson       sales   chicago
29         7900        clerk       sales   chicago
30         7676        staff  operations   chicago
31         7820  salesperson       sales    boston

### Задание 3
 Выбрать всех покупателей, с которыми работает продавец TURNER.

### Задание 4
Выбрать сотрудников, которые не закреплены за конкретным заказчиком.

### Задание 5
Выбрать имена продавцов, которые работают более чем с одним покупателем и названия покупателей, которых они обслуживают.

### Задание 6
Выбрать сумму всех продаж, которые обеспечил продавец TURNER.

### Задание 7
Выбрать сумму всех продаж товаров, связанных с теннисом, за лето 1990г.

### Задание 8
Выбрать имена и зарплаты самого старого и самого молодого сотрудника фирмы. ("Возраст" сотрудника определять по дате его поступления на работу в фирму).

### Задание 9
Для каждого продавца вывести разность между его зарплатой и средней зарплатой продавцов в фирме. Решение

### Задание 10
Вывести имена, средние инициалы и фамилии всех однофамильцев.

### Задание 11
Для каждого отдела выбрать процент, который составляют сотрудники отдела от общего числа сотрудников фирмы, и процент, который составляет зарплата отдела от общей зарплаты фирмы.

### Задание 12
Выбрать фамилии начальников тех отделов, которые имеют разветвленную структуру (т.е. тех начальников, у прямых подчиненных которых есть свои подчиненные).

### Задание 13
Выбрать названия отдела, заключившего сделки на самую большую сумму, и города, в котором он находится.

### Задание 14
Выбрать название товара, дату продажи, цену продажи для всех случаев, когда товары продавались ниже, чем за 75% их объявленной цены.

### Задание 15
Выбрать общую сумму скидок, предоставленных покупателю STADIUM SPORTS в 1989г.

### Задание 16
Выбрать список сотрудников фирмы с указанием фамилии непосредственного начальника каждого.

### Задание 17
 Для каждого сотрудника вывести разность между его зарплатой и средней зарплатой сотрудников, выполняющих те же функции.

### Задание 18
Вывести таблицу распределения объема продаж товара DUNK BASKETBALL INDOOR по годам.



### Задание 19
Выбрать фамилии тех сотрудников, у которых суммарный доход (зарплата + комиссионные) больше 2000.

### Задание 20
Выбрать названия товаров, для которых нынешняя цена увеличилась по сравнению с ценой на 15 декабря 1989г. более, чем на 15%.